drivers/qcommsimmodem/radio-settings.c

```diff
@@ -0,0 +1,269 @@
+	ofono_radio_settings_rat_mode_set_cb_t cb = cbd->cb;
+
+	if (message->error == RIL_E_SUCCESS) {
+		g_ril_print_response_no_args(rd->ril, message);
+		CALLBACK_WITH_SUCCESS(cb, cbd->data);
+	} else {
+		ofono_error("%s: rat mode setting failed", __func__);
+		CALLBACK_WITH_FAILURE(cb, cbd->data);
+	}
+}
+
+static void qcom_msim_set_2g_rat_cb(struct ril_msg *message,
+							gpointer user_data)
+{
+	struct qcom_msim_set_2g_rat *set_2g_rat_data = user_data;
+	struct ofono_radio_settings *rs = set_2g_rat_data->rs;
+	struct qcom_msim_pending_pref_setting *pps = set_2g_rat_data->pps;
+	struct radio_data *rd = ofono_radio_settings_get_data(rs);
+	ofono_radio_settings_rat_mode_set_cb_t cb;
+	struct parcel rilp;
+
+	pps->pending_gsm_pref_remaining -= 1;
+
+	if (message->error == RIL_E_SUCCESS) {
+		g_ril_print_response_no_args(rd->ril, message);
+		ofono_radio_settings_set_rat_mode(rs,
+						OFONO_RADIO_ACCESS_MODE_GSM);
+	} else {
+		ofono_error("%s: rat mode setting failed", __func__);
+		if (pps->cbd != NULL) {
+			cb = pps->cbd->cb;
+			CALLBACK_WITH_FAILURE(cb, pps->cbd->data);
+
+			g_free(pps->cbd);
+			pps->cbd = NULL;
+		}
+	}
+
+	if (pps->pending_gsm_pref_remaining == 0) {
+		if (pps->cbd != NULL) {
+			struct radio_data *pps_rd =
+					ofono_radio_settings_get_data(pps->rs);
+			g_ril_request_set_preferred_network_type(pps_rd->ril,
+							pps->pref, &rilp);
+
+			if (g_ril_send(pps_rd->ril,
+					RIL_REQUEST_SET_PREFERRED_NETWORK_TYPE,
+					&rilp, qcom_msim_set_rat_cb, pps->cbd,
+					g_free) == 0) {
+				ofono_error("%s: unable to set rat mode",
+								__func__);
+				cb = pps->cbd->cb;
+				CALLBACK_WITH_FAILURE(cb, pps->cbd->data);
+				g_free(pps->cbd);
+			}
+		}
+
+		g_free(pps);
+	}
+}
+
+static void qcom_msim_set_rat_mode(struct ofono_radio_settings *rs,
+			enum ofono_radio_access_mode mode,
+			ofono_radio_settings_rat_mode_set_cb_t cb,
+			void *data)
+{
+	struct radio_data *rd = ofono_radio_settings_get_data(rs);
+	struct cb_data *cbd = cb_data_new(cb, data, rs);
+	struct parcel rilp;
+	int pref = PREF_NET_TYPE_GSM_WCDMA;
+
+	switch (mode) {
+	case OFONO_RADIO_ACCESS_MODE_ANY:
+		pref = PREF_NET_TYPE_LTE_GSM_WCDMA;
+		break;
+	case OFONO_RADIO_ACCESS_MODE_GSM:
+		pref = PREF_NET_TYPE_GSM_ONLY;
+		break;
+	case OFONO_RADIO_ACCESS_MODE_UMTS:
+		pref = PREF_NET_TYPE_GSM_WCDMA;
+		break;
+	case OFONO_RADIO_ACCESS_MODE_LTE:
+		pref = PREF_NET_TYPE_LTE_GSM_WCDMA;
+		break;
+	}
+
+	if (pref != PREF_NET_TYPE_GSM_ONLY && multisim_rs_amount > 1) {
+		struct qcom_msim_pending_pref_setting *pps = g_try_new0(
+				struct qcom_msim_pending_pref_setting, 1);
+		int i;
+
+		pps->rs = rs;
+		pps->pref = pref;
+		pps->cbd = cbd;
+		pps->pending_gsm_pref_remaining = 0;
+
+		for (i = 0; i < QCOMMSIM_NUM_SLOTS_MAX; i++) {
+			struct radio_data *current_rd;
+			struct qcom_msim_set_2g_rat *set_2g_rat_data;
+
+			if (multisim_rs[i] == rs)
+				continue;
+
+			current_rd = ofono_radio_settings_get_data(
```

Comment:
You might want to change this to temp_rd, as it confused me because it's not really the current_rd ( ie. you return if multisim_rs[i] == rs ).  Maybe temp_rd would be a better name?